pkg/internal/owner/owner.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2020 The Bulward Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package owner
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// OwnerNameLabel references the name of the owner of this object.
+	OwnerNameLabel = "owner.bulward.io/name"
+	// OwnerNamespaceLabel references the namespace of the owner of this object.
+	OwnerNamespaceLabel = "owner.bulward.io/namespace"
+	// OwnerTypeLabel references the type of the owner of this object.
+	OwnerTypeLabel = "owner.bulward.io/type"
+)
+
+type generalizedListOption interface {
+	client.ListOption
+	client.DeleteAllOfOption
+}
+
+// SetOwnerReference sets a the owner as owner of object.
+func SetOwnerReference(owner, object runtime.Object, scheme *runtime.Scheme) (changed bool) {
+	objectAccessor, err := meta.Accessor(object)
+	if err != nil {
+		panic(fmt.Errorf("cannot get accessor for %T :%w", object, err))
+	}
+
+	labels := objectAccessor.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+
+	ownerLabels := labelsForOwner(owner, scheme)
+	for k, v := range ownerLabels {
+		if labels[k] != v {
+			changed = true
+		}
+		labels[k] = v
+	}
+	objectAccessor.SetLabels(labels)
+	return
+}
+
+// RemoveOwnerReference removes an owner from the given object.
+func RemoveOwnerReference(owner, object runtime.Object) (changed bool) {
```

Comment:
likewise, check that if you're removing owner reference that you're the owner (or the object is unowned); otherwise error out

Reply:
This function is copied from KubeCarrier directly, and it has not been used anywhere. let's keep it like this, if we needed it, we improve it later.